daemon/config/config.go

```diff
@@ -139,10 +159,11 @@ type TLSOptions struct {
 
 // DNSConfig defines the DNS configurations.
 type DNSConfig struct {
-	DNS           []net.IP `json:"dns,omitempty"`
-	DNSOptions    []string `json:"dns-opts,omitempty"`
-	DNSSearch     []string `json:"dns-search,omitempty"`
-	HostGatewayIP net.IP   `json:"host-gateway-ip,omitempty"`
+	DNS            []net.IP `json:"dns,omitempty"`
+	DNSOptions     []string `json:"dns-opts,omitempty"`
+	DNSSearch      []string `json:"dns-search,omitempty"`
+	HostGatewayIP  net.IP   `json:"host-gateway-ip,omitempty"` // Deprecated: this single-IP is migrated to HostGatewayIPs
+	HostGatewayIPs []string `json:"host-gateway-ips,omitempty"`
```

Comment:
Curious; was there a reason not to use `[]net.IP` for this? From some quick testing, marshalling / unmarshaling should work for that? See https://go.dev/play/p/m-9f_mVwuGW

```go
package main

import (
	"encoding/json"
	"net"
	"testing"
)

type config struct {
	HostGatewayIP  net.IP  `json:"host-gateway-ip,omitempty"` // Deprecated: this single-IP is migrated to HostGatewayIPs
	HostGatewayIPs []net.IP `json:"host-gateway-ips,omitempty"`
}

func TestGatewayIPS(t *testing.T) {
	cfg1 := config{
		HostGatewayIPs: []net.IP{
			net.ParseIP("2001:db8::1234"),
			net.ParseIP("192.0.2.1"),
		},
	}
	out, err := json.Marshal(cfg1)
	if err != nil {
		t.Fatal(err)
	}
	t.Log(string(out))

	const configJSON = `{"host-gateway-ips": ["2001:db8::1234", "192.0.2.1"]}`

	var cfg config
	err = json.Unmarshal([]byte(configJSON), &cfg)
	if err != nil {
		t.Fatal(err)
	}
	t.Logf("%+v\n", cfg.HostGatewayIPs)
}
```

Reply:
Because we don't have any separation between config read from the file, config from the command line, and the structs used to represent config internally.

A `[]string` can be used in `flags.Var(opts.NewNamedListOptsRef("host-gateway-ips", &conf.HostGatewayIPs, opts.ValidateIPAddress)` ... it'd be possible to write a `[]net.IP` (or, rather, `[]netip.Addr`) version of that. Perhaps it could also be rewritten to use generic types. But, I didn't do any of that! ... it wouldn't buy much in this case, the addresses are only used as strings, and they're validated as IP addresses.

Reply:
Ah, right, so there _is_ a `IPSliceVar` (and related) option in `spf13/pflag`. Haven't checked what's needed to wrap that into a "NamedListOpt" (or whatever it's named) though; https://github.com/moby/moby/blob/803534fb4e784d3e6bef9846d64ce4960ad1ec80/vendor/github.com/spf13/pflag/ip_slice.go#L151-L160

Reply:
It doesn't quite fit together like that, `IPSliceVar` takes a `*[]net.IP`, which can't be a `NamedOption` (so it can't implement the equivalence between command-line `--host-gateway-ip` and config file `"host-gateway-ips"`).

So, I've added a `NamedIPListOpts`, which stores values in a `[]netip.Addr` and implements interface `NamedOption`.

I added it as a third commit to make it easier to see the change ... how's it looking?